### ISSUE TITLE
DROP TABLE statements can't run in a transaction block

### ISSUE THREAD

**Issue by samay-sharma** [citusdata/citus#774]
``` sql
postgres=# begin;
BEGIN
postgres=# drop table table1;
ERROR:  DROP distributed table cannot run inside a transaction block
CONTEXT:  SQL statement "SELECT master_drop_all_shards(v_obj.objid, v_obj.schema_name, v_obj.object_name)"
PL/pgSQL function citus_drop_trigger() line 39 at PERFORM
```

**Comment by samay-sharma** [comment]
Feedback from a user: "Our drop table commands in our migration scripts do not work inside of a transaction scope, which makes perfect sense why it’s not supported. However we have no control over our migration tool, and all DB migration frameworks and ORMs I’ve ever used wrap migrations inside of transactions always. So it would be nice if we could “turn off” Citus’ protection layer here and proceed with the distributed DDL even if we’re breaking out of the transaction’s semantics."

**Comment by ozgune** [comment]
@lfittl and @samay-sharma -- I had a question on this issue as it relates to #668.

How likely do you think is it that users will look to run ``DROP TABLE`` in a transaction block?

@jasonmp85 -- What was our reason for closing the other issue and not this one? Is it because we handle ``DROP TABLE`` through event triggers?

If so, we now know that ORMs commonly run ``TRUNCATE`` on individual tables to implement unit tests. @lfittl / @mtuncer -- are we good with our implementation in #86?

**Comment by lfittl** [comment]
@ozgune For this particular issue I don't think its required for Rails/ActiveRecord, and in particular [database_cleaner](https://github.com/DatabaseCleaner/database_cleaner) which is often used in unit/integration tests.

Whilst database_cleaner has an option to drop tables and then re-create them, but typically I've seen it being used with truncation.

There is another case where a DROP TABLE will occur in a transaction, and that is inside Rails database migrations, in particular when rolling back (since migrations always run in a transaction by default). Its not a major issue, but something a user might notice.

I'd say in general this issue here is a nice to have in regards to ORM support (from a Rails/AR focus), not a must have.

**Comment by lfittl** [comment]
Some clarifications after talking through this with @ozgune:

If we don't support DROP TABLE in transactions (i.e. as it is today), some ORM frameworks will return an error when executing a "down migration", i.e. a database change that one wants to roll back.

Note that this is also something @samay-sharma noted in #668: https://github.com/citusdata/citus/issues/668#issuecomment-245280975

Right now you need to disable transactions, like here: https://github.com/citusdata/citus-example-ad-analytics/blob/master/db/migrate/20160622202229_add_impression_click_count_rollups.rb#L31

**Comment by metdos** [comment]
Apparently, this issue is fixed by @byucesoy in 6.1 release: https://github.com/citusdata/citus/commit/d80e7849a4b22cdd0fb17e0a380c156fade7e701
